charon/tests/generate-ml.rs

```diff
@@ -0,0 +1,467 @@
+                                    fields.iter().map(|f| f.name.as_ref().unwrap()).join("; ");
+                                format!("`List [ {pat} ]")
+                            }
+                        } else {
+                            // Struct variant
+                            let pat = fields
+                                .iter()
+                                .map(|f| {
+                                    let name = f.name.as_ref().unwrap();
+                                    let var = if f.is_opaque() { "_" } else { name };
+                                    format!("(\"{name}\", {var});")
+                                })
+                                .join(" ");
+                            format!("`Assoc [ {pat} ]")
+                        };
+                        let pat = format!("`Assoc [ (\"{name}\", {inner_pat}) ]");
+                        let construct_fields =
+                            fields.iter().map(|f| f.name.as_ref().unwrap()).join(", ");
+                        let construct = format!("{rename} ({construct_fields})");
+                        build_branch(ctx, &pat, &fields, &construct)
+                    }
+                })
+                .join("\n")
+        }
+        TypeDeclKind::Opaque => todo!(),
+        TypeDeclKind::Error(_) => todo!(),
+    };
+    build_function(ctx, decl, &branches)
+}
+
+#[test]
+// Test is ignored by default. See the top for how to run it.
+#[ignore]
+fn generate_ml() -> Result<()> {
+    let charon_llbc = "tests/generated-ml/charon-itself.llbc";
+    const RUN_CHARON: bool = true;
+    if RUN_CHARON {
+        // Call charon on itself
+        let mut cmd = Command::cargo_bin("charon")?;
+        cmd.arg("--cargo-arg=--lib");
+        cmd.arg("--errors-as-warnings");
+        cmd.arg("--dest-file");
+        cmd.arg(charon_llbc);
+        let output = cmd.output()?;
+
+        if !output.status.success() {
+            let stderr = String::from_utf8(output.stderr.clone())?;
+            bail!("Compilation failed: {stderr}")
+        }
+    }
+
+    let crate_data: CrateData = {
+        let file = File::open(charon_llbc)?;
+        let reader = BufReader::new(file);
+        serde_json::from_reader(reader)?
+    };
+
+    let mut ctx = GenerateCtx {
+        crate_data,
+        contains_raw_span: Default::default(),
+    };
+
+    let mut name_to_type: HashMap<String, &TypeDecl> = Default::default();
+    for ty in &ctx.crate_data.types {
+        let long_name = repr_name(&ctx.crate_data, &ty.item_meta.name);
+        if long_name.starts_with("charon_lib") {
+            let short_name = ty.item_meta.name.name.last().unwrap().as_ident().0.clone();
+            name_to_type.insert(short_name, ty);
+        }
+        name_to_type.insert(long_name, ty);
+    }
+    let raw_span = name_to_type.get("RawSpan").unwrap();
+    ctx.contains_raw_span = contains_id(&ctx.crate_data, raw_span.def_id);
+
+    let names: &[&[&str]] = &[
```

Comment:
Would be good to add a comment here to explain that we will extract those groups of definitions in the order listed below.